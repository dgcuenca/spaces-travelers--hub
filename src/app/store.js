import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import rocketReducer from '../features/rockets/rockets.slice';
import missionReducer from '../features/missions/missions.slice';

const store = configureStore({
  reducer: {
    rockets: rocketReducer,
    missions: missionReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  devTools: process.env.NODE_ENV !== 'production',
});
export default store;
