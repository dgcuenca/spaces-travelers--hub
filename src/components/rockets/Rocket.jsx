import React from 'react';
import PropTypes from 'prop-types';
import styles from './Rocket.module.scss';

const Rocket = ({ rocket }) => (
  <div className={styles.rocket}>
    <div className={styles.rocket__left}>
      <img
        className={styles.rocket__img}
        src={rocket.flickr_image}
        alt={rocket.name}
      />
    </div>
    <div className={styles.rocket__right}>
      <h2 className={styles.rocket__title}>{rocket.name}</h2>
      <p className={styles.rocket__description}>{rocket.description}</p>
      <button
        type="button"
        className={styles.rocket__button}
      >
        Reserve Rocket
      </button>
    </div>
  </div>
);

Rocket.defaultProps = {
  rocket: null,
};

Rocket.propTypes = {
  rocket: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    flickr_image: PropTypes.string.isRequired,
  }),
};

export default Rocket;
