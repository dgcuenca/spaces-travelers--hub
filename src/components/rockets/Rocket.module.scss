.rocket {
  display: flex;
  flex-direction: row;
  gap: 1rem;
  position: relative;

  &__left {
    position: relative;
    width: 33%;
  }

  &__right {
    width: 70%;
    display: flex;
    flex-direction: column;
    gap: 0.625rem;
  }

  &__img {
    height: 180px;
    width: 100%;
  }

  &__title {
    font-size: 2rem;
  }

  &__description {
    font-size: 0.875rem;
    letter-spacing: 2%;
    line-height: 1.25rem;
  }

  &__button {
    font-size: 1rem;
    cursor: pointer;
    letter-spacing: 1%;
    padding: 0.625rem 0.813rem;
    width: fit-content;
    background-color: #157ffb;
    color: #fff;
    border-radius: 5px;
  }

  &__button--cancel {
    border: 1px solid #6d757d;
    color: #6d757d;
    font-size: 1rem;
    cursor: pointer;
    letter-spacing: 1%;
    padding: 0.625rem 0.813rem;
    width: fit-content;
    border-radius: 5px;
  }
}
