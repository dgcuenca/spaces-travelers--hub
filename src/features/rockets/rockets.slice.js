import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const ROCKET_URL = 'https://api.spacexdata.com/v4/rockets';

export const fetchRockets = createAsyncThunk(
  'rockets/FETCH_ROCKETS',
  async () => {
    const response = await axios.get(ROCKET_URL);
    const rockets = response.data;
    const modifiedRockets = rockets.map((rocket) => ({
      id: rocket.id,
      name: rocket.name,
      type: rocket.name,
      description: rocket.description,
      flickr_image: rocket.flickr_images[0],
    }));
    return modifiedRockets;
  },
);

const initialState = {
  rockets: [],
  status: 'idle',
  error: null,
};

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    reserveRocket: (state, action) => {
      const rockets = state.rockets.map((rocket) => {
        if (rocket.id !== action.payload) return rocket;
        if (rocket.reserved) {
          return {
            id: rocket.id,
            name: rocket.name,
            type: rocket.name,
            description: rocket.description,
            flickr_image: rocket.flickr_image,
          };
        }
        return { ...rocket, reserved: true };
      });
      return { ...state, rockets };
    },
    // cancelReservation: (state, action) => {
    //   const rockets = state.rockets.map((rocket) => {
    //     if (rocket.id === action.payload) {
    //       return {
    //         id: rocket.id,
    //         name: rocket.name,
    //         type: rocket.name,
    //         description: rocket.description,
    //         flickr_image: rocket.flickr_image,
    //       };
    //     }
    //     return { ...rocket, reserved: true };
    //   });
    //   return { ...state, rockets };
    // },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchRockets.pending, (state) => {
        const newState = { ...state, status: 'loading' };
        return newState;
      })
      .addCase(fetchRockets.fulfilled, (state, action) => {
        const newState = {
          ...state,
          status: 'succeeded',
          rockets: action.payload,
        };
        return newState;
      })
      .addCase(fetchRockets.rejected, (state, action) => {
        const newState = {
          ...state,
          status: 'failed',
          error: action.error.message,
        };
        return newState;
      });
  },
});

export const getAllRockets = (state) => state.rockets.rockets;
export const getRocketsStatus = (state) => state.rockets.status;
export const getRocketsError = (state) => state.rockets.error;
export const reservedRockets = (state) => {
  const filteredRockets = state.rockets.rockets.filter(
    (rocket) => rocket.reserved === true,
  );
  const reduceRockets = filteredRockets.map((rocket) => {
    const { id, name } = rocket;
    const modifiedRocket = {
      id,
      name,
    };
    return modifiedRocket;
  });
  return reduceRockets;
};

export const { reserveRocket } = rocketsSlice.actions;

export default rocketsSlice.reducer;
