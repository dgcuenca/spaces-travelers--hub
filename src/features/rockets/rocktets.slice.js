import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const ROCKET_URL = 'https://api.spacexdata.com/v3/rockets';

export const fetchRockets = createAsyncThunk(
  'rockets/FETCH_ROCKETS',
  async () => {
    const response = await axios.get(ROCKET_URL);
    const rockets = response.data;
    const modifiedRockets = rockets.map((rocket) => ({
      id: rocket.id,
      name: rocket.rocket_name,
      type: rocket.rocket_type,
      description: rocket.description,
      flickr_image: rocket.flickr_images[0],
    }));
    return modifiedRockets;
  },
);

const initialState = {
  rockets: [],
  status: 'idle',
  error: null,
};

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchRockets.pending, (state) => {
        const newState = { ...state, status: 'loading' };
        return newState;
      })
      .addCase(fetchRockets.fulfilled, (state, action) => {
        const newState = {
          ...state,
          status: 'succeeded',
          rockets: action.payload,
        };
        return newState;
      })
      .addCase(fetchRockets.rejected, (state, action) => {
        const newState = {
          ...state,
          status: 'failed',
          error: action.error.message,
        };
        return newState;
      });
  },
});

export default rocketsSlice.reducer;
