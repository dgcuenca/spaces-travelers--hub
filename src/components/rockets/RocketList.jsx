import React from 'react';
import PropTypes from 'prop-types';
import Rocket from './Rocket';
import styles from './RocketList.module.scss';

const RocketList = ({ rockets }) => (
  <div className={styles.rocketList}>
    {rockets.map((rocket) => (
      <Rocket
        key={rocket.id}
        rocket={rocket}
      />
    ))}
  </div>
);

RocketList.defaultProps = {
  rockets: [],
};

RocketList.propTypes = {
  rockets: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
    }),
  ),
};

export default RocketList;
