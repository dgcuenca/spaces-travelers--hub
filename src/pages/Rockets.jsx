import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  fetchRockets,
  getRocketsError,
  getRocketsStatus,
  getAllRockets,
} from '../features/rockets/rockets.slice';
import RocketList from '../components/rockets/RocketList';
import styles from './Rockets.module.scss';

const Rockets = () => {
  const dispatch = useDispatch();
  const rocketStatus = useSelector(getRocketsStatus);
  const rockets = useSelector(getAllRockets);
  const error = useSelector(getRocketsError);
  useEffect(() => {
    if (rocketStatus === 'idle') {
      dispatch(fetchRockets());
    }
  }, [rocketStatus, dispatch]);

  let contentToDisplay = '';
  if (rocketStatus === 'loading') {
    contentToDisplay = <div className={styles.rockets__status}>Loading...</div>;
  } else if (rocketStatus === 'succeeded') {
    contentToDisplay = <RocketList rockets={rockets} />;
  } else if (rocketStatus === 'failed') {
    contentToDisplay = (
      <em className={styles['rockets__status--failed']}>{error}</em>
    );
  }

  return <section className={styles.rockets}>{contentToDisplay}</section>;
};

export default Rockets;
