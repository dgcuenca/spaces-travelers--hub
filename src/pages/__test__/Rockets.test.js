/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import Rockets from '../Rockets';
import store from '../../app/store';
import { fetchRockets } from '../../features/rockets/rockets.slice';

describe('Rockets Page', () => {
  test('renders Rockets Page', async () => {
    await store.dispatch(fetchRockets());
    const tree = render(
      <Provider store={store}>
        <BrowserRouter>
          <Rockets />
        </BrowserRouter>
      </Provider>,
    );
    expect(tree.toJSON).toMatchSnapshot();
  });
});
