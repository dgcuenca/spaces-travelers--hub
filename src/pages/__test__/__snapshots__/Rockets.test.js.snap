// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets Page renders Rockets Page 1`] = `undefined`;
